### ISSUE TITLE
Impersonation in asp.net core

### ISSUE THREAD

**Issue by coding786** [MarimerLLC/cslaforum#696]
As asp.net core does not implement impersonation, how csla work arounds with the same if using asp.net core based app server ?

in wcf based app server, impersonation was possible at server side dataportal configuration. How to do the same in asp.net core based app server ?

**Version and Platform**
CSLA version: 4.6.9
OS: Windows
Platform: WPF, ASP.NET Core

**Comment by itgoran** [comment]
@coding786 
Depending on context (for what purposes you need it) I've created following 'Helper' method to do that..

```
public class ExecuteAs
{
	public static void User(int userId, string userFullName, Action action)
	{
            var context = IoC.Container.Resolve<IHttpContextAccessor>().HttpContext;
            var oldUserPrincipal = context.User as CustomApplicationUserPrincipal;
            var oldExecutionInfoBase = oldUserPrincipal.ExecutionContextInformation;          
            context.User = newUserPrincipal;
            action();            
            context.User = oldUserPrincipal;
	}
}
```
Note that ```Action action``` is used to pass action which you wish to execute. 
Note that code is not CSLA related it's just way how to execute an action with additional user (our actions are CSLA commands)

**Comment by rockfordlhotka** [comment]
Maybe this is helpful?

https://nicolgit.github.io/cross-micro-services-authentication-Azure-B2C/

**Comment by coding786** [comment]
Thanks ! In the context of csla.net , what configuration settings should be done for impersonation ? 
We used to enable impersonation in wcf channel factory in wpf client proxy object earlier. Not sure what all settings to be done on wpf client(using HTTP Proxy) when using core based app server ?

**Comment by rockfordlhotka** [comment]
Kind of what you asked in #695?

I'm not 100% sure, but I suspect you'll need to follow the guidance in the Using CSLA 4 book for custom authentication. If I understand correctly, ASP.NET Core doesn't allow for the user's client-side identity (on a client Windows workstation) to flow to the server.

That said, I also suspect that _if you are hosting ASP.NET Core in IIS_ that it WILL work as expected, because the impersonation will occur at the IIS level, before ASP.NET Core is even invoked. If that's true (and I suspect it is) then you can use the normal guidance for configuring the data portal to use configuration.

Really CSLA doesn't do any of this work by the way. When you configure CSLA all you are doing is telling CSLA how it should act _assuming you've configuring the underlying platform_ correctly.

In other words, if you tell CSLA to use Windows authentication, all that means is that CSLA _assumes_ that you've configured your web server to flow the user's identity from the client workstation to the server (via IIS or ASP.NET classic). To be precise, it means CSLA will not serialize the client-side principal/identity to the server, because it assumes Windows has done this on your behalf.

If you tell CSLA to use custom authentication, all that means is that CSLA will serialize the client-side principal/identity to the server, because it assumes Windows has _not_ done this on your behalf.

**Comment by coding786** [comment]
Thanks ! I am getting below unauthorized error on a DataPortal.Fetch in 3 tier from wpf client to core app portal ,anything if missing, pls suggest 
On WPF client have set AuthenticationType to Windows
On Server side - have configured IIS for Windows Auth & disabled anonymous through  IIS manager at website authentication
I am explicitly setting in startup of asp.net core
       Csla.ApplicationContext.AuthenticationType="Windows"
       Csla.ApplicationContext.User = new WindowsPrincipal(WindowsIdentity.GetCurrent());

System.Net.HTTP.HttpRequestException : Response status code does not indicate success: 401(unauthorized)
Error is coming from dataportal.cs class

**Comment by rockfordlhotka** [comment]
If you set the CSLA authentication mode to 'Windows' that tells CSLA not to serialize the client-side principal/identity to the server.

That means that you are allowing Windows itself to do the impersonation. And as you have set IIS to require Windows authn that should be correct.

What I don't know, and so can't help you, is whether you should be setting the current user to a new `WindowsPrincipal`? Is that the recommended solution from some Microsoft docs?
